enhancements/node-pull-credentials/pull-credentials.md

```diff
@@ -0,0 +1,236 @@
+  methods of providing them to different parts of the codebase).
+
+## Proposal
+
+### User Stories
+
+As an OpenShift user
+I want to be able to import imagestreams from any image registry that the
+cluster nodes can pull images from
+so that I can use them without manually creating any extra credentials.
+
+As an OpenShift user
+I want to be able to use, during builds, images from any image registry that
+the cluster nodes can pull images from
+so that I can use these images as base for my own images.
+
+As an OpenShift user
+I want to be able to pull-through images from any registry that the cluster
+nodes can pull images from
+so that I can use image registry cache to speed-up my builds.
+
+### Implementation Details/Notes/Constraints
+
+This is the node's filesystem path we are taking into account:
+
+- `/var/lib/kubelet/config.json` contains the node's pull secret.
+
+#### Image Stream Import
+
+- Mount the pull secret as `readOnly`, `hostPathreadOnly` in
+  `openshift-apiserver` deployment under `/node/var/lib/kubelet/config.json`.
+
+```yaml
+volumeMounts:
+- mountPath: /node/var/lib/kubelet/config.json
+  name: node-pull-credentials
+  readOnly: true
+volumes:
+- name: node-pull-credentials
+  hostPath:
+    path: /var/lib/kubelet/config.json
+    type: File
+```
+
+- When a user imports an imagestream, `openshift-apiserver` will merge all pull
+  credentials found in `/node/var/lib/kubelet/config.json` with other
+  credentials that may exist in the namespace.
+
+
+#### Builds
+
+- `controller-manager` would mount pull credentials inside `build` pod as 
+  `hostPath`(similar to what is done for Image Stream Import).
+- Builder image parses node's pull credentials and uses them during build,
+  merging with other pull credentials that are linked to the `builder` service
+  account. If the `BuildConfig` specifies a pull secret, we will continue the
+  current behavior of using the provided pull secret as an override.
+
+#### Registry pull-through
+
+- As done for Image Stream Import, mount pull credentials inside the image
+  registry pod.
+- Pull credentials will then be consumed by the registry.
+
+### Risks and Mitigations
+
+#### Pull credentials for registry may exist on namespace and on node
+
+User may have created a secret containing credentials for a registry that node
+already has credentials to.
+
+Mitigations:
+
+Always prioritize namespace secrets over node's credentials. If a credential
+exists on the namespace, do not use node's credentials.
+
+#### Pull credentials inside build pod may be visible to the user
+
+Cluster wide credentials mounted inside the builder pod may be a security risk
+as the user may be able to shell into the pod and copy them. Another possible
+risk is that the user may copy the credentials into a resulting image.
+
+Mitigations:
+
+As far as I verified(and this needs to be once more tested) it is impossible
+to spawn a shell inside the builder pods. I also tried to copy the credentials
```

Comment:
Ephemeral pods will allow sharing process namespace and thus provide you with access to it. That's a major problem.

Reply:
See https://kubernetes.io/docs/concepts/workloads/pods/ephemeral-containers/

Reply:
This issue also applies to "openshift-apiserver" during image stream import or is something only for builds?

Reply:
@ricardomaraschini what @soltysh is saying is that even though we block oc rsh into a build-pod, someone (may) be able to create an ephemeral container in the build-pod's process namespace and use that mechanism to get into it instead.

@deads2k are there security mechanisms that would prevent someone from abusing this?  much like we don't allow you to exec/rsh into a pod you can't "create" yourself (so normal users can't exec/rsh into a privileged container, as i understand it), does the ephemeral container api have similar restrictions?

Reply:
@bparees there is no limitation at this time to my knowledge.  its not clear to me that we could use rbac appropriately in a namespace to block the subresource to create the ephemeral container as build pods would be in the same namespace as other pods we would want to allow ephemeral access.

Reply:
that subresource effectively updates the pod resource.  If i can't create privileged pods, i would not expect to be able to edit prvileged pods and i'd expect that restriction to extend to subresources that modify the pod.  No?

Reply:
spoke with @derekwaynecarr offline.  SCC will need to handle this (ensure that a user who can't create a particular pod also cannot invoke this subresource which edits the pod).

between @openshift/sig-developer-experience and @openshift/sig-master we need to ensure this is covered by the time the ephemeral container api goes beta (at which point it will be on by default).

Reply:
I have enabled EphemeralContainers on kube-apiserver and simulated the feature, as `kubeadmin` I have been able to patch the pod normally as expected but as a regular user I  received the following message back:

```
$ oc replace --raw /api/v1/namespaces/build/pods/img-1-build/ephemeralcontainers -f ./ec.json 
Error from server (Forbidden): pods "img-1-build" is forbidden: User "user" cannot update resource "pods/ephemeralcontainers" in API group "" in the namespace "build"
$
```

ec.yaml
```
$ cat ec.json 
{
    "apiVersion": "v1",
    "kind": "EphemeralContainers",
    "metadata": {
            "name": "img-1-build"
    },
    "ephemeralContainers": [{
        "command": [
            "sh"
        ],
        "image": "busybox",
        "imagePullPolicy": "IfNotPresent",
        "name": "debugger",
        "stdin": true,
        "tty": true,
        "terminationMessagePolicy": "File"
    }]
}
```